### DISCUSSION TITLE
Losing reactivity after an action - autocomplete component

### DISCUSSION THREAD

**Discussion by Ashraam** [livewire/livewire#4646] [1. Help]
Hi, I'm trying to make an autocomplete field, it works great until I "unselect" a city.

When I type a city name or zip code, it shows the list. When I select a city in the list it selects the city, then when I click on the "unselect" action, it unselects and show back the search field but all reactivity is lost. I can type into the field but the component  does nothing (no JS error nor Server error)

What do I miss here ? Is there any tricks to enhance this component ?

Here is the component markup

```
<div class="position-relative w-100">
    @if($city)
        <div class="input-group">
            <input type="text" class="form-control form-control-solid border" value="{{ $city->nom }} ({{ $city->code_postal }})" readonly />
            <span class="input-group-text border cursor-pointer" wire:click="unselect">
                <i class="fas fa-times fs-4"></i>
            </span>
        </div>
    @else
        <input class="form-control form-control-solid" placeholder="Rechercher une ville par son nom ou code postal...." wire:model.debounce="query" wire:keydown.escape="cancel" />
        
        @if(strlen($query) >= 3)
        <x-card class="w-100 z-10 position-absolute">
            <x-card-body>
                @forelse($results as $city)
                <button class="d-block w-100 btn btn-sm btn-active-light-primary text-start" wire:click="select({{ $city->id }})">{{ $city->nom }} ({{ $city->code_postal }})</button>
                @empty
                <div class="text-warning text-center fs-6"><i class="fa text-warning fa-exclamation-triangle"></i> Aucun résultat</div>
                @endforelse
            </x-card-body>
        </x-card>
        @endif
    @endif
    
    <input type="hidden" name="city_id" value="{{ optional($city)->id }}" />
</div>
```

and here is the component himself

```
<?php

namespace App\Http\Livewire;

use App\Models\City;
use Livewire\Component;

class SearchCities extends Component
{
    public $query;

    public $results = [];

    public $city;

    public function updatedQuery($value)
    {
        if (strlen($value) >= 3) {
            $this->search();
        } else {
            $this->results = [];
        }
    }

    public function search()
    {
        $this->results = City::where('code_postal', 'like', $this->query . '%')->orWhere('commune', 'like', '%' . $this->query . '%')->orWhere('lieu_dit', 'like', '%' . $this->query . '%')->limit(15)->get();
    }

    public function select($id)
    {
        $this->city = $this->results->find($id);

        $this->cancel();
    }

    public function unselect()
    {
        $this->reset('city');
    }

    public function cancel()
    {
        $this->reset(['query', 'results']);
    }

    public function mount($id = null)
    {
        if ($id) {
            $this->city = City::find($id);
        }
    }

    public function render()
    {
        return view('livewire.search-cities');
    }
}
```

Thank you

![Capture d’écran 2022-02-10 à 11 34 13](https://user-images.githubusercontent.com/4686383/153389405-7c82edb2-6f5e-4485-9151-93faf45218b7.png)

**Comment by silnarm** [comment]
Hi, this might be a dom-diffing issue, the majority of the content in the blade is in an @if @else @endif, and I think the DOM diffing algo is getting confused on the unselect action thinking the input it now sees (with wire:model) is actually the previous input it already had (the readonly, unwired one) rather than an entirely different one.
Stick a key on the wired input and see if the problem goes away:
`<input ... wire:model.debounce="query" wire:keydown.escape="cancel" wire:key="foo" />
`

**Comment by joshhanley** [comment]
@Ashraam yeah @silnarm is correct, that is probably a morphdom issue. You should be making use of `wire:key` when using elements inside `@if/@else` blocks as well as in a `@foreach` loop, which instructs morphdom on how to handle the element changes. See here for details https://laravel-livewire.com/docs/2.x/troubleshooting#dom-diffing-issues
